### PR TITLE
Cleanup interstitial logic

### PR DESCRIPTION
**Pull request by nikosdouvlis**
## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [x] `@clerk/nextjs`
- [x] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/clerk-expo`
- [x] `@clerk/backend-core`
- [x] `@clerk/clerk-sdk-node`
- [x] `@clerk/edge`
- [ ] `build/tooling/chore`

## Description

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.

- Split the `getAuthState` logic into 2 distinct scenarios, `handleRequestWithTokenInHeader` and `handleRequestWithTokenInCookie` that can be handled separately
- For the token in cookie scenario, the complex interstitial logic has been extracted into `InterstitialRules`. These rules run sequentially and in the same order as before
- To further simplify `getAuthState`, it no longer fetches the interstitial. This responsibility is moved back to the caller

### CODE REVIEW

File: packages/backend-core/src/Base.ts [hunk 1]
```diff
@@ -88,14 +90,13 @@ export class Base {
...
-   * @param {string} token Clerk JWT verification token
+   * @param jwt Clerk JWT verification token
    * Modify the RSA public key from the Clerk PEM supplied and return a contructed CryptoKey.
    * You will find that at your application dashboard (https://dashboard.clerk.dev) under Settings ->  API keys
    *
    */
-  loadCryptoKey = async (key?: string): Promise<CryptoKey> => {
-    if (!key) {
+  loadCryptoKey = async (jwt?: string): Promise<CryptoKey> => {
+    if (!jwt) {
```
Comment (igneel64) [inline packages/backend-core/src/Base.ts:99]: This is actually a key here, how did the change came about ?

**Comment by igneel64** [comment]
I personally don't prefer the nested logic versus the line by line as it was before. It makes the flowchart easier to follow in my view.
I am not sure how this came about but I would like to know how other people feel about it.

File: packages/backend-core/src/Base.ts (hunk 1 above)
Comment (nikosdouvlis) [reply packages/backend-core/src/Base.ts:99]: Probably from the comment - but it should be `key`!

File: packages/backend-core/src/util/interstitialRules.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+// prevent interstitial rendering
+export const nonBrowserRequestInDevRule: InterstitialRule = (params, key) => {
+  if (isDevelopmentOrStaging(key) && !params.userAgent?.startsWith('Mozilla/')) {
+    return { status: AuthStatus.SignedOut, errorReason: AuthErrorReason.HeaderMissingNonBrowser };
+  }
+  return undefined;
+};
+
+// Is this correct? In cross-origin requests the use of Authorization header is mandatory
+export const crossOriginRequestWithoutCors: InterstitialRule = params => {
```
Comment (colinclerk) [inline packages/backend-core/src/util/interstitialRules.ts:31]:
```suggestion
export const crossOriginRequestWithoutHeader: InterstitialRule = params => {
```

**Review by colinclerk — APPROVED** [review]
I didn't review the refactor very significantly, just the names of the rules.

In general I like having the verbose rule names listed here, though I commented a lot on them to add extra clarity.  Let me know if there's any confusion.

This is a little bit of a scary change. We'd benefit from having an external test suite that we could use to run tests against this SDK and others written for other languages, so we know they're always behaving consistently.

**Comment by SokratisVidros** [comment]
Team what's the status of this?

**Comment by nikosdouvlis** [comment]
Closing this one for now, will tackle it as it's own task during the true SSR project.
Some of the safest changes have been extracted to a separate PR we can merge now: https://github.com/clerkinc/javascript/pull/233

Ticket: https://www.notion.so/clerkdev/Refactor-interstitial-logic-856b68a9f76845a3b0e5ddd1c4f3a727

**Pull request closed without merging**